### PR TITLE
Multinomial

### PR DESCRIPTION
**Pull request by tongxin**
This multinomial is a Triton conversion from the Pytorch counterpart. 

Performance figure updated.
```
benchmark/test_special_perf.py Operator multinomial Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.411968            0.244064
6144                  0.690048            0.449952
11264                 0.931296             0.53248
16384                  1.13818            0.664704
21504                  1.23549            0.677472
26624                  1.39875            0.745216
31744                  1.54067             0.80096
36864                  1.74266            0.934592
41984                   1.8904            0.965376
47104                  2.07027             1.03923
52224                  2.23818             1.08048
57344                  2.45094             1.19658
62464                  2.59354             1.19482
67584                  2.71466             1.29584
72704                  2.92397              1.3409
77824                  3.11814             1.42378
Operator multinomial Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.451392            0.267328
6144                   0.80832            0.466176
11264                  1.08074            0.544896
16384                  1.33142            0.743872
21504                  1.45101            0.778528
26624                  1.65642            0.899328
31744                  1.85642            0.982752
36864                  2.11466             1.14086
41984                  2.31005             1.20314
47104                  2.52592             1.30464
52224                  2.71603             1.37322
57344                  2.94445             1.50336
62464                  3.13027             1.54714
67584                  3.29763             1.67389
72704                  3.50435             1.76499
77824                  3.69504             1.85344
.
```

### CODE REVIEW

File: tests/test_special_ops.py [hunk 1]
```diff
@@ -215,3 +215,30 @@ def test_accuracy_resolve_conj(shape, dtype):
...
+    dist = torch.zeros(size=shape, dtype=dtype, device="cuda")
+    Index = [5, 13, 42]
+    dist[..., Index] = 1
+    with flag_gems.use_gems():
+        res_out = torch.multinomial(dist, n_samples, True)
+    assert torch.all(torch.isin(res_out, torch.tensor(Index, device="cuda")))
+
+
+@pytest.mark.parametrize("pool", [100, 2048])
+@pytest.mark.parametrize("dtype", [torch.float32])
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: does multinomial without replacement support torch.float16 and torch.bfloat16?

File: benchmark/test_special_perf.py [hunk 2]
```diff
@@ -56,3 +56,20 @@ def resolve_conj_arg(dtype, batch, size):
...
+        n_samples = 10000
+        return (dist, n_samples, True)
+
+    bench = Benchmark(
+        op_name="multinomial",
+        torch_op=torch.multinomial,
+        arg_func=multinomial_args,
+        dtypes=(torch.float16, torch.float32),
+        batch=POINTWISE_BATCH,
+        sizes=(100, 1024, 30000),
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py]: suggest passing 5 sizes so that tests on multinomial could align with other operators

File: src/flag_gems/ops/multinomial.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Sampling without replacement
+    if (not with_replacement) or n_samples == 1:
+        # In case of with_replacement, sampling is approximated by selecing
+        # the top k indices over sorted probabilities with an exponential pertubation
+        # s = argmax( p / q ) where q ~ Exp(1)
+        q = torch.empty_like(prob).exponential_(1.0)
+        s = torch.div(prob, q, out=q)
+        if n_samples == 1:
+            return torch.argmax(s, dim=-1)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/multinomial.py]: theoretically, the only sample may also be other elements besides the largest one.

File: tests/test_distribution_ops.py [hunk 4]
```diff
@@ -36,3 +38,50 @@ def test_accuracy_exponential_(shape, dtype):
...
+    Index = [5, 13, 42]
+    dist[..., Index] = 1
+    with flag_gems.use_gems():
+        res_out = torch.multinomial(dist, n_samples, True)
+    index, tally = torch.unique(res_out, sorted=True, return_counts=True)
+    assert index.tolist() == Index
+    # Do a simple Chi-square test
+    tally = np.array(tally.tolist())
+    expected = tally
+    expected[:] = tally.mean()
```
Comment (StrongSpoon) [inline tests/test_distribution_ops.py]: why is the mean number expected?

**Review by StrongSpoon — COMMENTED** [review]
I don't understand why chi-square proves the accuracy.

File: benchmark/test_special_perf.py (hunk 2 above)
Comment (tongxin) [reply benchmark/test_special_perf.py]: Sure.

File: src/flag_gems/ops/multinomial.py (hunk 3 above)
Comment (tongxin) [reply src/flag_gems/ops/multinomial.py]: The original probabilities are already modified by a random factor so argmax may change too.

File: tests/test_distribution_ops.py (hunk 4 above)
Comment (tongxin) [reply tests/test_distribution_ops.py]: Here I devised a test where the three indices are sampled uniformly with replacement and then test the uniformity of the output with a Chi-square test.  This is equivalent to testing samples of a categorical distribution. To use Chi-square. we need to supply the actual sample counts and the expected sample counts. Since the original probabilities are evenly divided between the three indices the expected sample count for each category is just the mean.

File: tests/test_special_ops.py (hunk 1 above)
Comment (tongxin) [reply tests/test_special_ops.py]: I'll add these two.

File: tests/test_distribution_ops.py (hunk 4 above)
Comment (tongxin) [reply tests/test_distribution_ops.py]: As to why using Chi-square to test accuracy, the short answer is Chi-square is quite often used verifying if a large number of samples are drawn from a categorical distribution. The multinomial function in Pytorch is no other than sampling from a given categorical distribution.

File: src/flag_gems/ops/multinomial.py (hunk 3 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/multinomial.py]: got it

File: tests/test_distribution_ops.py (hunk 4 above)
Comment (StrongSpoon) [reply tests/test_distribution_ops.py]: got it

**Comment by tongxin** [comment]
Also added fused_norm_cumsum for better perf.

File: benchmark/test_special_perf.py [hunk 5]
```diff
@@ -73,3 +73,20 @@ def resolve_conj_arg(dtype, batch, size):
...
+    def multinomial_args(dtype, batch, size):
+        dist = torch.rand((batch, size), dtype=dtype, device="cuda")
+        n_samples = 10000
+        return (dist, n_samples, True)
+
+    bench = Benchmark(
+        op_name="multinomial",
+        torch_op=torch.multinomial,
+        arg_func=multinomial_args,
+        dtypes=(torch.float16, torch.float32),
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py:111]: why not run benchmark on bf16?

File: tests/test_special_ops.py [hunk 6]
```diff
@@ -254,3 +256,40 @@ def test_accuracy_resolve_conj(shape, dtype):
...
         z = y.resolve_conj()
     assert not z.is_conj()
+
+
+@pytest.mark.parametrize("shape", UT_SHAPES_1D + UT_SHAPES_2D)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+@pytest.mark.parametrize("n_samples", [1000])
+def test_accuracy_multinomial_with_replacement(shape, dtype, n_samples):
+    if shape[-1] == 1:
+        dist = torch.rand(size=shape, dtype=dtype, device="cuda")
```
Comment (StrongSpoon) [inline tests/test_special_ops.py:356]: tests on cpu need to_reference

File: tests/test_special_ops.py [hunk 7]
```diff
@@ -254,3 +256,40 @@ def test_accuracy_resolve_conj(shape, dtype):
...
+                res_out = torch.multinomial(dist, n_samples, True)
+            res_dist = torch.gather(dist, -1, res_out)
+            assert torch.all(res_dist)
+
+
+@pytest.mark.parametrize("pool", UT_SHAPES_1D + UT_SHAPES_2D)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_multinomial_without_replacement(pool, dtype):
+    n_draws = 10
+    dist = torch.rand(size=(pool,), dtype=dtype, device="cuda").broadcast_to(
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: ditto

File: tests/test_special_ops.py [hunk 8]
```diff
@@ -254,3 +256,40 @@ def test_accuracy_resolve_conj(shape, dtype):
...
+        n_draws, pool
+    )
+    n_samples = pool
+    with flag_gems.use_gems():
+        res_out = torch.multinomial(dist, n_samples, False)
+    # Verifies uniqueness
+    sorted_samples, _ = res_out.sort(dim=1)
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: for 1d shapes, I think the sort dim should be -1.

File: tests/test_special_ops.py [hunk 9]
```diff
@@ -254,3 +256,40 @@ def test_accuracy_resolve_conj(shape, dtype):
...
+    assert torch.all(sorted_samples == torch.arange(pool, device="cuda"))
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: gems_assert_equal could be the same

**Review by StrongSpoon — COMMENTED** [review]
finished

File: benchmark/test_special_perf.py (hunk 5 above)
Comment (tongxin) [reply benchmark/test_special_perf.py:111]: Bfloat is not supported yet for torch.multinomial.

File: tests/test_special_ops.py (hunk 6 above)
Comment (tongxin) [reply tests/test_special_ops.py:356]: In this test, we don't need a reference result. It's device only test.

File: tests/test_special_ops.py (hunk 7 above)
Comment (tongxin) [reply tests/test_special_ops.py]: Same here.

File: tests/test_special_ops.py (hunk 8 above)
Comment (tongxin) [reply tests/test_special_ops.py]: In this test, the input dist and output samples are in 2d.

Comment (tongxin) [reply tests/test_special_ops.py]: but dim=-1 is better indeed.

File: src/flag_gems/ops/cumsum.py [hunk 10]
```diff
@@ -76,3 +76,40 @@ def cumsum(inp, dim=1, *, dtype=None):
...
+    y = tl.cumsum(x, 0)
+    # tl.store(tmp_out + tl.program_id(0) + row_start + row_off, y, mask=row_off == (BLOCK - 1), cache_modifier='.cg')
+    # y_sum = tl.load(tmp_out + tl.program_id(0) + row_start + BLOCK - 1, cache_modifier='.cg')
+    y = y / y_sum
+    tl.store(out + row_start + row_off, y, mask=row_off < K)
+
+
+def fused_renorm_cumsum(inp, dim=-1):
+    logging.debug("GEMS RENORM_CUMSUM")
+    assert inp.dtype in (torch.float16, torch.float32, torch.float64)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/cumsum.py]: allow inp.dtype to be torch.bfloat16

Comment (tongxin) [reply src/flag_gems/ops/cumsum.py]: Sure.

**Review by StrongSpoon — APPROVED** [review]
lgtm
